clang/docs/checkedc/rfc/Include-Checked-Hdrs.md

```diff
@@ -0,0 +1,182 @@
+       that contains `foo_checked.h`. In this new file, we add the following:
+
+             #ifdef IMPLICIT_INCLUDE_CHECKED_HDRS
+             #include <foo_checked.h>
+             #else
+             #include_next <foo.h>
+             #endif
+     - In `foo_checked.h`, we modify `#include <foo.h>` to
+       `#include_next <foo.h>`.
+
+     - The way this solution will work:
+        - If a program includes `foo.h`, clang will pick up either
+          `foo_checked.h` or the system `foo.h` depending on whether or not
+          `-DIMPLICIT_INCLUDE_CHECKED_HDRS` is specified on the compilation
+          commandline.  Therefore specifying this flag will cause the implicit
+          inclusion of the checked counterpart of `foo.h`, which will make it
+          convenient for automation. Not specifying the commandline flag will
+          not perform implicit inclusion, providing a mechanism to opt out of 
+          implicit inclusion. Infinite recursion is avoided because of
+	  `#include_next`.
+        - If a program includes `foo_checked.h` current behavior will prevail.
+          The above flag has no effect on the explicit inclusion of checked
+          header files.
+
+     - **Pros**: 1) The solution provides a way to opt into implicit inclusion
+       of checked header files. 2) The existing approach of explicit inclusion
+       is supported. 3) In most cases integration with the build system is easy:
+       the flag `-DIMPLICIT_INCLUDE_CHECKED_HDRS` is passed to the compilation
+       of all source files through the `CFLAGS` variable. 4) No changes are
+       required to the order in which include directories are searched. 5) It
+       is convenient for automation.
+
+     - **Cons**: 1) In some cases integration with the build system will be more
+       involved if the above flag needs to be passed to the compilation of most
+       source files and avoided for a few source files that may want to directly
+       include system header files in order to avoid Checked-C-specific
+       declarations. 2) It is difficult to accomodate clang-specific
+       declarations.
+
+## Proposed Solution:
+
+The proposed solution is as follows:
+  - We will implicitly include checked header files by default and provide the 
+    compilation flag NO_IMPLICIT_INCLUDE_CHECKED_HDRS to opt out of the implicit
+    inclusion.
+  - For header files that do not have clang-specific declarations, but have
+    Checked-C-specific declarations (there are 13 such header files at present),
+    we will do the following:
+      - For each system header file, say `foo.h`, that does not have
+        clang-specific declarations but has Checked-C-specific declarations, we
+        will add a new file also called `foo.h` that will contain the following:
+
+            #if !defined __checkedc || defined NO_IMPLICIT_INCLUDE_CHECKED_HDRS
+            #include_next <foo.h>
+            #else
+            #include <foo_checked.h>
+            #endif
```

Comment:
I realize this can be substantially simplified if you don't mind some header files being read multiple times.  They probably are or have to be guarded anyway, so the only cost I see is performance, unless there is a concern about a header file indirectly including itself recursively even if the recursion is immediately stopped by a guard.
```suggestion
            #include_next <foo.h>
            #if defined __checkedc && !defined NO_IMPLICIT_INCLUDE_CHECKED_HDRS
            #include <foo_checked.h>
            #endif
```

Reply:
I understand that this suggestion has the advantage of a uniform approach for all header files but it relies on guard flags to avoid recursion. In order to keep the solution simple and maintainable, we do prefer to avoid a header file including itself recursively without relying on guard flags.
Additionally, in the implementation, there are more flags entering into the fray in order to handle the absence of some posix header files like `unistd.h`, `arpa/inet.h` and `sys/socket.h` in the Windows environment. This makes it even more necessary to keep the solution simple and also comprehensible.

Reply:
When I think about this now, I think I got a little carried away with my proposal and you were right to decline it.  The resulting code looks simpler on the surface, but it's significantly more complicated to verify that it works correctly.

FWIW, an alternative way to achieve a uniform approach is to have each header `foo.h` with no Clang-specific declarations use your existing design for headers with Clang-specific declarations, effectively as if Clang already had a `foo.h` containing nothing but `#include_next <foo.h>` followed by an empty set of Clang-specific declarations.  You could decide whether the resulting short, boilerplate `foo.h` goes in the `checkedc-clang` or `checkedc` repository; the former would achieve total uniformity.  But this would increase the number of header files you have to maintain by about 50% (in addition to `foo.h` and `foo_checked.h`, you would have `foo_checked_internal.h`) for minimal practical benefit, so I imagine it's not worth it.  (The only practical benefit I see is that if the user had some problem with the system `foo.h` but had another way to get the plain-C declarations they want, they could then get the Checked C declarations by including `foo_checked_internal.h` with no risk of pulling in the system `foo.h`.  But I think this would be very unusual, and the user could use the hacky workaround of manually defining the header guard for `foo.h` before including `foo_checked.h`.)

Reply:
Yes, we did earlier consider having the uniform approach of creating `*_checked_internal.h` for all header files that have Checked-C-specific declarations. At present, there are 14 such header files of which `inttypes.h` is the only file with clang-specific declarations. We decided not to create `*_checked_internal.h` for all because 13 of these files will be redundant at present.
If a user wants custom plain-C declarations and also Checked C declarations for, say, `foo.h`, under the present system he/she can 1) include either `foo.h` (for implicit inclusion) or `foo_checked.h` (for explicit inclusion) and 2) add `-nostdlibinc` and the include directory containing custom plain-C declarations on the compilation command line.